Tiered sale with rewards from promo codes

friday-james opened the pull request.

friday-james committed:
tiered sale

friday-james committed:
tier-ed sale with bonus percentage

friday-james committed:
tier-ed sale with more test cases

friday-james committed:
add documentation

friday-james committed:
add saveCashAll functin

friday-james committed:
add helper functions to halt, unhalt and updateWhitelist

friday-james committed:
add comments

friday-james committed:
more tests

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -0,0 +1,271 @@
+        uint8 bonusPercentage;  // Additional bonus percentage for this tier
+        bytes32 whitelistRootHash;
+        bool isHalt;
+        bool isIntegerSale;
```
can we add, allowPromoCode here? and if it is false, no promo code / discount will be added

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -0,0 +1,271 @@
+
+        uint256 totalCost = _amount * _price;  // in gwei
+        uint256 baseOwnerPercentage = totalCost * 8 / 100;
+        uint256 masterOwnerPercentage = totalCost * 2 / 100;
```
I think this 8 and 2 can be modified for future sales, so can we make it as a constant and modifiable as well

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -0,0 +1,271 @@
+        emit Cash(_msgSender(), withdrawAmount, 0);
+    }
+
+    function _validatePromoCode(string memory _promoCode) internal view {
```
where is this being called again?

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -0,0 +1,271 @@
+            tokenSold += purchasedAmountPerTier[tierIds[i]][promoCodeAddress];
+        }
+        require(tokenSold > 0, "Promo code owner has not purchased any token");
+        // if the promo code is an address, check if it has purchased any node
```
PS: This will be 5% discount and 5% to the wallet address

calebwick replied:
and can we add a flag where this is disabled/enabled?

friday-james replied:
if the code is an address, there'll be a 5% discount to the purchaser and a 5% reward to the referrer wallet address?

calebwick replied:
correct

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -0,0 +1,271 @@
+        }
+    }
+
+    function updateWhitelist(string memory _tierId, bytes32 _whitelistRootHash) public onlyOperator {
```
can we make one for maxTotalPurchasable too, i think having a separate updater for maxTotal and publicAllo, and isHalt is safer

friday-james committed:
code review

friday-james committed:
validate promo code on claim

friday-james committed:
remove IFTieredSale1

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -122,6 +129,7 @@ contract IFTieredSale is ReentrancyGuard, AccessControl, IFFundable, IFWhitelist
 
 
     function addPromoCode(string memory _code, uint8 _discountPercentage, address _promoCodeOwnerAddress, address _masterOwnerAddress) public onlyOperator {
+        require(allowPromoCode, "Promo code is not allowed")
```
I saw that allowPromoCode is on each Tier? Or do we have a global variable?

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -150,6 +159,19 @@ contract IFTieredSale is ReentrancyGuard, AccessControl, IFFundable, IFWhitelist
+            require(MerkleProof.verify(_merkleProof, whitelistRootHash, keccak256(abi.encodePacked(msg.sender))), "Invalid proof");
+            require(purchasedAmountPerTier[_tierId][msg.sender] + _amount <= _allocation, "Purchase exceeds allocation");
+        }
+        executePurchase(_tierId, _amount, tiers[_tierId].price, _promoCode);
```
why does this still use _promoCode? Won't the test fail?

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -208,8 +230,25 @@ contract IFTieredSale is ReentrancyGuard, AccessControl, IFFundable, IFWhitelist
+
+    function updateRewards(string memory _tierId, uint8 _baseOwnerPercentage, uint8 _masterOwnerPercentage) public onlyOperator {
+        tiers[_tierId].baseOwnerPercentage = _baseOwnerPercentage;
+        tiers[_tierId].masterOwnerPercentage = _masterOwnerPercentage;
```
Don't think this is on the tier level right?

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -168,13 +190,13 @@ contract IFTieredSale is ReentrancyGuard, AccessControl, IFFundable, IFWhitelist
-        promoCodes[_promoCode].promoCodeOwnerEarnings += baseOwnerPercentage + bonus;
-        promoCodes[_promoCode].masterOwnerEarnings += masterOwnerPercentage;
+        totalRewardsUnclaimed += baseOwnerRewards + masterOwnerRewards + bonus;
+        promoCodes[_promoCode].promoCodeOwnerEarnings += baseOwnerRewards + bonus;
```
Can we skip this if promoCode is not provided?

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -35,6 +41,7 @@ contract IFTieredSale is ReentrancyGuard, AccessControl, IFFundable, IFWhitelist
         bytes32 whitelistRootHash;
         bool isHalt;
         bool isIntegerSale;
+        bool allowPromoCode;
```
also add allowWalletPromoCode, if false, the promo code using wallet should not happen

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -0,0 +1,311 @@
+        emit Cash(_msgSender(), withdrawAmount, 0);
+    }
+
+    function _validatePromoCode(string memory _promoCode) internal view {
```
why is this not called on purchasing step?

friday-james replied:
becoz the promo code can be invalid on purchase but valid on withdrawal
if the promo code is an address and the address purchases a node at the end of the sale, the promo code rewards shd be valid

calebwick replied:
for this one, i might prefer to reject the tx if the wallet has not bought. Because if using this, there might be some edge cases

friday-james committed:
modify tier and promo code settings

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -190,13 +206,30 @@ contract IFTieredSale is ReentrancyGuard, AccessControl, IFFundable, IFWhitelist
+                promoCodes[_promoCode].promoCodeOwnerAddress = promoCodeAddress;
+            }
+            promoCodes[_promoCode].promoCodeOwnerEarnings += totalCost * addressPromoCodePercentage / 100;
+            promoCodes[_promoCode].totalPurchased += totalCost;
```
Is this intended to not add totalRewardsUnclaimed?

friday-james replied:
oh right, shd add it here

friday-james committed:
validate address promo code on purchase

calebwick commented on contracts/IFTieredSale.sol:
```diff
@@ -0,0 +1,361 @@
+    // State variables
+    mapping(string => Tier) public tiers;
+    mapping(string => mapping(address => uint256)) public purchasedAmountPerTier;  // tierId => address => amount in ether
+    mapping(string => uint256) public codePurchaseAmount;  // promo code => total purchased amount in ether
```
codePurchaseAmount this is not being used anywhere

friday-james committed:
increment variables

friday-james committed:
fix checkTierWhitelist

friday-james committed:
not to block purchase when maxAllocationPerWallet is 0

friday-james committed:
more doc and handle case when variables are 0

friday-james committed:
add view functions and allow batch withdraw reward

friday-james committed:
check promo code owner by msg.sender

friday-james committed:
add test cases

friday-james committed:
handle the case where the owner is both base and master

friday-james committed:
bump solidity compiler version

friday-james committed:
flexible reward percentage for each promo code

friday-james committed:
validate promo code setting

friday-james committed:
Update MAX_BONUS_PERCENTAGE

friday-james committed:
make sure promo code is not already exist

friday-james committed:
Audit feedback:
- correct codePurchaseAmount calculation
- remove unused code
- load promoCodeAddress accurately
- add sanity checks
- add setter function

friday-james committed:
fix promo code address parsing

friday-james committed:
decouple wallet promo code and regular promo code logic

friday-james committed:
correct variable name

friday-james committed:
decouple wallet reward withdrawal logic

calebwick committed:
Update license

friday-james committed:
reject purchase with own wallet address promo code

friday-james committed:
Merge branch 'tiered-sale' of github.com:ImpossibleFinance/impossible-node-sale into tiered-sale
